### PR TITLE
Improve error msg for CCS request on node without remote cluster role

### LINKED ISSUE #59683: CCS IndexNotFound Exception when disable remote_connect
Reproduced both on 7.3.1 and 7.8.0.

According documentation:
```
node.remote_cluster_client
By default, any node in the cluster can act as a cross-cluster client and connect to remote clusters. The node.remote_cluster_client setting can be set to false (defaults to true) to prevent certain nodes from connecting to remote clusters. Remote cluster requests must be sent to a node that is allowed to act as a cross-cluster client.
```
When configure some nodes, like data node, set `node.remote_cluster_client` to `false` in 7.8 or set `cluster.remote.connect` to `false`  in 7.3, the remote clusters was not working as expected, and if you run query against the remote cluster, it will randomly get `index_not_found_exception`, as below:
CCS-roundrobin-index-not-found

seems those remote requests are still routing to these nodes and this doesn't work as expected, these nodes should be skipped i guess.

But if you check the log, it says `password authentication failed for elastic`,
Jietu20200714-213301

And also this issue happens in proxy mode as well.

Another side effect on Kibana:
- Monitoring page refused to display, says permission error.
- Remote clusters page lost/disappear the data.

### PR DESCRIPTION
**Pull request by dnhatn**
If a CCS request executes on a node without the remote cluster client role, then it will fail with IndexNotFoundException. The error message is quite misleading. We should instead report that the node does not have the remote cluster client role instead.

Closes #59683

### CODE REVIEW

**Comment by javanna** [comment]
I think that this change deserves more discussion. From the docs: 

> By default, any node in the cluster can act as a cross-cluster client and connect to remote clusters. The node.remote_cluster_client setting can be set to false (defaults to true) to prevent certain nodes from connecting to remote clusters. Remote cluster requests must be sent to a node that is allowed to act as a cross-cluster client.

Adding automatic rerouting is a new feature rather than a bugfix, as the current behaviour is by design. What is not desirable is how requests end up failing today: we should rather return a clear message consistently, which I seem to recall we did in the past but maybe it got lost in some refactoring.

**Comment by dnhatn** [comment]
@javanna @jimczi I've updated this PR to improve the error message as we discussed. Would you please take a look? Thank you!

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java [hunk 1]
```diff
@@ -73,6 +79,10 @@ protected RemoteClusterAware(Settings settings) {
         for (String index : requestIndices) {
             int i = index.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR);
             if (i >= 0) {
+                if (isRemoteClusterClientEnabled == false) {
+                    assert remoteClusterNames.isEmpty() : remoteClusterNames;
+                    throw new IllegalStateException("node [" + nodeName + "] does not have the remote cluster client role enabled");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java:84]: I am not too sure about `IllegalStateException` here. It's more like a user error as they should rather send the request to another node? Maybe `IllegalArgumentException` would be more appropriate but I am not 100% happy with it either

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSRemoteClusterClientRoleIT.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        final String nodeWithoutRemoteClusterClientRole = localCluster.startNode(NodeRoles.onlyRole(DiscoveryNodeRole.DATA_ROLE));
+        final IllegalStateException error = expectThrows(IllegalStateException.class, () ->
+            localCluster.client(nodeWithoutRemoteClusterClientRole)
+                .prepareSearch("demo", "cluster_a:prod")
+                .setQuery(new MatchAllQueryBuilder())
+                .setAllowPartialSearchResults(false)
+                .setSize(1000)
+                .get());
+        assertThat(error.getMessage(),
+            equalTo("node [" + nodeWithoutRemoteClusterClientRole + "] does not have the remote cluster client role enabled"));
```
Comment (javanna) [inline server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSRemoteClusterClientRoleIT.java:73]: would it be possible to verify what status code would get returned to users at the REST layer?

**Review by javanna — COMMENTED** [review]
I left a couple of comments. My main concern is that I don't think this can be backported to 7.x as there can still be local indices with `:` in their names that were created in 6.x.

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java]: Yes, I think we should consider it a bad request. I've adjusted it in https://github.com/elastic/elasticsearch/pull/60351/commits/78ba655192c7788b0082da15526f0ba0bcc79691.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSRemoteClusterClientRoleIT.java (hunk 2 above)
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSRemoteClusterClientRoleIT.java]: I've strengthened the assertion in https://github.com/elastic/elasticsearch/pull/60351/commits/78ba655192c7788b0082da15526f0ba0bcc79691. I don't know how to verify the rest status in this IT test. Any suggestion?

**Comment by dnhatn** [comment]
> My main concern is that I don't think this can be backported to 7.x as there can still be local indices with : in their names that were created in 6.x.

Let's merge this in 8.x. I will open the fix for 7.x separately.

@javanna Thanks for review. Can you take another look?

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterServiceTests.java [hunk 3]
```diff
@@ -218,6 +221,18 @@ public void testGroupIndices() throws IOException {
         }
     }
 
+    public void testGroupIndicesWithoutRemoteClusterClientRole() throws Exception {
+        final Settings settings = onlyRoles(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "node-1").build(),
+            Sets.newHashSet(DiscoveryNodeRole.DATA_ROLE));
+        try (RemoteClusterService service = new RemoteClusterService(settings, null)) {
+            assertFalse(service.isEnabled());
+            assertFalse(service.isCrossClusterSearchEnabled());
+            final IllegalStateException error = expectThrows(IllegalStateException.class,
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/transport/RemoteClusterServiceTests.java:230]: this needs updating?

**Review by javanna — APPROVED** [review]
LGTM besides the comment I left, I re-labelled to target 8.0 only.

**Comment by dnhatn** [comment]
Thanks Luca!

**Pull request merged**